Marlin/src/module/stepper.cpp

```diff
@@ -194,6 +194,8 @@ uint32_t Stepper::acceleration_time, Stepper::deceleration_time;
   uint8_t Stepper::steps_per_isr = 1; // Count of steps to perform per Stepper ISR call
 #endif
 
+hal_timer_t Stepper::time_spent_in_isr = 0, Stepper::time_spent_out_isr = 0;
```

Comment:
I'm thinking of making a (hidden) option to turn this off and use the existing timing method, in the unlikely event that there is some edge case we haven't seen yet. One of the things that occurred to me was a potential situation where we do the timing calculation for a set of small identical segments, and as a result of reducing the multi-stepping based on the timing of one segment, the `time_spent_in_isr` goes down just enough to make the next segment calculation favor more multi-steps, which makes `time_spent_in_isr` go up just enough to make the next segment calculation favor fewer multi-steps, which makes `time_spent_in_isr` go down just enough to make the next segment calculation favor more multi-steps….

Is that something that could actually occur with just the right combination of segment-length and feedrate?

Reply:
I agree with the option to select the old behaviour. As you say, who knows what edge case there might be out there.

On the specific question about the behviour continually flipping between more and fewer multi-steps: If flipping like this happened, it could occur every time `block_phase_isr()` runs. However when multi-stepping doubles the amount of time required by the ISR does not decrease by as much as half. And since the ISR needs to consume less than 50% of the processor time before multi-stepping will halve again, I am pretty sure there is sufficient hysteresis to prevent continual flipping.